src/realm_dart_logger.cpp

```diff
@@ -0,0 +1,113 @@
+////////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+#include <sstream>
+#include <set>
+#include <mutex>
+#include <thread>
+#include <map>
+#include <chrono>
+
+#include "realm_dart_logger.h"
+#include <realm/object-store/c_api/types.hpp>
+
+auto& dart_logger_mutex = *new std::mutex;
+bool is_core_logger_callback_set = false;
+std::map<Dart_Port, realm_log_level_e> dart_send_ports;
+realm_log_level_e default_log_level = RLM_LOG_LEVEL_INFO;
+
+
+bool isPortRegistered(Dart_Port key)
+{
+    return (dart_send_ports.find(key) != dart_send_ports.end());
+}
+
+realm_log_level_e calucale_minimum_log_level()
+{
+    auto min_log_level = default_log_level;
+    if (dart_send_ports.size() > 0)
+    {
+        auto min_element = std::min_element(dart_send_ports.begin(), dart_send_ports.end(),
+            [](std::pair<Dart_Port, realm_log_level_e> const& prev, std::pair<Dart_Port, realm_log_level_e> const& next) {
+            return prev.second < next.second;
+        });
+        min_log_level = min_element->second;
+    }
+    return min_log_level;
+}
+
+RLM_API void realm_dart_release_logger(Dart_Port port) {
+    std::lock_guard lock(dart_logger_mutex);
+    if (isPortRegistered(port))
+    {
+        dart_send_ports.erase(port);
+        auto minimum_level = calucale_minimum_log_level();
+        realm_set_log_level(minimum_level);
+    }
+}
+
+bool send_message_to_scheduler(Dart_Port port, realm_log_level_e level, const char* message)
+{
+    Dart_CObject c_level;
+    c_level.type = Dart_CObject_kInt32;
+    c_level.value.as_int32 = level;
+
+    Dart_CObject c_message;
+    c_message.type = Dart_CObject_kString;
+    c_message.value.as_string = (char*)message;
```

Comment:
We should use `const_cast` here

Reply:
Also this seems very unsafe are we sure we are copying the memory from this pointer?

Reply:
I changed it to `const_cast`, but I don't understand what you mean with the question.